lib/common/tests/rules/pcmk__evaluate_attr_expression_test.c

```diff
@@ -0,0 +1,718 @@
+{
+    assert_attr_expression(EXPR_TYPE_VERSION_FAILS, pcmk_rc_before_range);
+}
+
+/*
+ * Test PCMK_XA_OPERATION
+ */
+
+#define EXPR_OP_MISSING                                 \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_ATTRIBUTE "='foo' "                     \
+        PCMK_XA_VALUE "='bar' />"
+
+static void
+op_missing(void **state)
+{
+    assert_attr_expression(EXPR_OP_MISSING, pcmk_rc_unpack_error);
+}
+
+#define EXPR_OP_INVALID                                 \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_ATTRIBUTE "='foo' "                     \
+        PCMK_XA_OPERATION "='not-an-operation' "        \
+        PCMK_XA_VALUE "='bar' />"
+
+static void
+op_invalid(void **state)
+{
+    assert_attr_expression(EXPR_OP_INVALID, pcmk_rc_unpack_error);
+}
+
+#define EXPR_OP_LT_PASSES                               \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_LT "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='20' />"
+
+static void
+op_lt_passes(void **state)
+{
+    assert_attr_expression(EXPR_OP_LT_PASSES, pcmk_rc_ok);
+}
+
+#define EXPR_OP_LT_FAILS                                \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_LT "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='2' />"
+
+static void
+op_lt_fails(void **state)
+{
+    assert_attr_expression(EXPR_OP_LT_FAILS, pcmk_rc_after_range);
+}
+
+#define EXPR_OP_GT_PASSES                               \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_GT "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='2' />"
+
+static void
+op_gt_passes(void **state)
+{
+    assert_attr_expression(EXPR_OP_GT_PASSES, pcmk_rc_ok);
+}
+
+#define EXPR_OP_GT_FAILS                                \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_GT "' "       \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='20' />"
+
+static void
+op_gt_fails(void **state)
+{
+    assert_attr_expression(EXPR_OP_GT_FAILS, pcmk_rc_before_range);
+}
+
+#define EXPR_OP_LTE_PASSES                              \
+        "<" PCMK_XE_EXPRESSION " " PCMK_XA_ID "='e' "   \
+        PCMK_XA_TYPE "='" PCMK_VALUE_INTEGER "' "       \
+        PCMK_XA_OPERATION "='" PCMK_VALUE_LTE "' "      \
+        PCMK_XA_ATTRIBUTE "='num' "                     \
+        PCMK_XA_VALUE "='10' />"
+
+static void
+op_lte_passes(void **state)
```

Comment:
Make sure LTE and GTE pass when the attribute is less than and greater than the value, respectively. Right now we're only testing the "attribute equals value" case.

Reply:
might as well